### PR TITLE
feat: add onComplete param to popWhile function in RouterExt.kt

### LINKED ISSUE #98: Add onComplete callback to Router.popWhile
We can pass data from one child to another e.g. when pushing / replacing / bringingToFront we can provide necessary data from configuration class or tell a child to execute a function, but to be consistent with `pop(...)` function, an `onComplete: (isSuccess: Boolean) -> Unit` callback may be added.
Passing data from a child to its indirect parent is a popular use case, there're many applications where we might need to skip some (or even all) configurations e.g. until the root element is shown.

### PR DESCRIPTION
**Pull request by Tommyten**
Closes #98

### CODE REVIEW

File: decompose/src/commonMain/kotlin/com/arkivanov/decompose/router/RouterExt.kt [hunk 1]
```diff
@@ -32,9 +32,19 @@ fun <C : Any> Router<C, *>.pop(onComplete: (isSuccess: Boolean) -> Unit = {}) {
 
 /**
  * Drops the configurations at the top of the stack while the [predicate] returns true
+ *
+ * @param onComplete called when the navigation is finished (either synchronously or asynchronously).
+ * The `isSuccess` argument is `true` if the stack size was greater than 1 and a component was popped,
```
Comment (arkivanov) [inline decompose/src/commonMain/kotlin/com/arkivanov/decompose/router/RouterExt.kt]: This statement is not quite correct for this case. Would it be possible to change this to something like the following?

> The `isSuccess` argument is `true` if at least one component has been popped.

**Review by arkivanov — CHANGES_REQUESTED** [review]

**Comment by arkivanov** [comment]
Ahh, and could you please run `./gradlew apiDump` and commit the changes?

**Comment by Tommyten** [comment]
Thank you for the suggestions, I've updated my branch

**Comment by arkivanov** [comment]
Sorry, the sample master-detail doesn't compile. It's better to also allow the most common use case without `onComplete` callback: `router.popWhile {}`.

Perhaps we need two variants - the existing one as it is now, and the new one with all required arguments:

`router.popWhile(predicate)` and `router.popWhile(predicate, onComplete)`.

Do you want to update? Or alternatively I can do it in your branch. If you will update, please don't forget to rerun `apiDump` task at the end.

**Comment by arkivanov** [comment]
So the existing function will delegate to the new one, similar to `Router.navigate` - https://github.com/arkivanov/Decompose/blob/96d08a3f0d8e67e36e466109b6b9a02e72dd1c67/decompose/src/commonMain/kotlin/com/arkivanov/decompose/router/RouterExt.kt#L9

**Comment by Tommyten** [comment]
I will do it as soon as I get to my PC.
Sorry for the hassle 😅
I'm still getting used to the project but hoping to be involved more soon!

**Comment by arkivanov** [comment]
Thanks for your contributions!

> Sorry for the hassle

I should have noticed this at the very beginning. :-)

**Comment by Tommyten** [comment]
How about we just swap the parameters to be
```kotlin
inline fun <C : Any> Router<C, *>.popWhile(
    crossinline onComplete: (isSuccess: Boolean) -> Unit = {},
    crossinline predicate: (C) -> Boolean
) {
    navigate(
        transformer = { stack -> stack.dropLastWhile(predicate) },
        onComplete = { newStack, oldStack -> onComplete(newStack.size < oldStack.size) }
    )
}
```

This way the API would be backwards compatible. Although having the first parameter being optional is not always best practice, we could prevent having to overload `popWhile`. You call.

**Comment by arkivanov** [comment]
> How about we just swap the parameters to be
> 
> ```kotlin
> inline fun <C : Any> Router<C, *>.popWhile(
>     crossinline onComplete: (isSuccess: Boolean) -> Unit = {},
>     crossinline predicate: (C) -> Boolean
> ) {
>     navigate(
>         transformer = { stack -> stack.dropLastWhile(predicate) },
>         onComplete = { newStack, oldStack -> onComplete(newStack.size < oldStack.size) }
>     )
> }
> ```
> 
> This way the API would be backwards compatible. Although having the first parameter being optional is not always best practice, we could prevent having to overload `popWhile`. You call.

There are a few points:

- I believe this change is still binary incompatible, it's only source compatible
- When using both arguments, it's looks intuitive to handle completion in the lambda outside parentheses
- There is already `Navigator.navigate(transformer, onComplete)` method, and an extension `Navigator.navigate(transformer)`. This was done like this for the reason mentioned above. Would be good to keep `popWhile` aligned.

**Comment by arkivanov** [comment]
There is a build error:

> API check failed for project decompose.

Could you please run `./gradlew :decompose:apiDump` and commit the changes?

**Review by arkivanov — APPROVED** [review]

**Comment by arkivanov** [comment]
Thanks for the contribution!

**Pull request merged**